cabal-install/src/Distribution/Client/ProjectConfig/Legacy.hs

```diff
@@ -223,26 +226,46 @@ instantiateProjectConfigSkeletonWithCompiler os arch impl _flags skel = go $ map
 
-projectSkeletonImports :: ProjectConfigSkeleton -> [ProjectConfigImport]
+projectSkeletonImports :: ProjectConfigSkeleton -> [ProjectConfigPath]
 projectSkeletonImports = view traverseCondTreeC
 
-parseProjectSkeleton :: FilePath -> HttpTransport -> Verbosity -> [ProjectConfigImport] -> FilePath -> BS.ByteString -> IO (ParseResult ProjectConfigSkeleton)
-parseProjectSkeleton cacheDir httpTransport verbosity seenImports source bs = (sanityWalkPCS False =<<) <$> liftPR (go []) (ParseUtils.readFields bs)
+-- | Parses a project from its root config file, typically cabal.project.
+parseProject :: FilePath -> FilePath -> HttpTransport -> Verbosity -> ProjectConfigToParse -> IO (ParseResult ProjectConfigSkeleton)
+parseProject rootConfig = parseProjectSkeleton [] "" (ProjectConfigPath $ rootConfig :| [])
+
+-- | Parses project configuration recursively, following imports.
+parseProjectSkeleton :: [ProjectConfigPath] -> FilePath -> ProjectConfigPath -> FilePath -> HttpTransport -> Verbosity -> ProjectConfigToParse -> IO (ParseResult ProjectConfigSkeleton)
+parseProjectSkeleton [] _ (ProjectConfigPath (rootPath :| [])) cacheDir httpTransport verbosity configToParse =
+  let (projectDir, projectFileName) = splitFileName rootPath; projectPath = ProjectConfigPath $ projectFileName :| []
+   in parseProjectSkeleton [projectPath] projectDir projectPath cacheDir httpTransport verbosity configToParse
+parseProjectSkeleton seenImports dir rootOrImport cacheDir httpTransport verbosity (ProjectConfigToParse bs) =
+  (sanityWalkPCS False =<<) <$> liftPR (go rootOrImport []) (ParseUtils.readFields bs)
   where
-    go :: [ParseUtils.Field] -> [ParseUtils.Field] -> IO (ParseResult ProjectConfigSkeleton)
-    go acc (x : xs) = case x of
-      (ParseUtils.F l "import" importLoc) ->
-        if importLoc `elem` seenImports
-          then pure . parseFail $ ParseUtils.FromString ("cyclical import of " ++ importLoc) (Just l)
+    go :: ProjectConfigPath -> [ParseUtils.Field] -> [ParseUtils.Field] -> IO (ParseResult ProjectConfigSkeleton)
+    go configPath acc (x : xs) = case x of
+      (ParseUtils.F l "import" importLoc) -> do
+        let importLocPath = ProjectConfigPath (importLoc <| coerce configPath)
+        let fullLocPath = fullConfigPathRoot dir importLocPath
+        normLocPath <- canonicalizeConfigPath dir importLocPath
+        normSeenImports <- nub <$> mapM (canonicalizeConfigPath dir) seenImports
+        info verbosity $ "import location, relative: " ++ show importLoc
+        info verbosity $ "import path, relative:\n" ++ showProjectConfigPath importLocPath
+        info verbosity $ "import path, normalized:\n" ++ showProjectConfigPath normLocPath
+        info verbosity "seen imports:\n"
+        mapM_ (info verbosity . showProjectConfigPath) normSeenImports
+        if (lengthConfigPath (nubConfigPath normLocPath) < lengthConfigPath normLocPath) || normLocPath `elem` normSeenImports
```

Comment:
I've explained this in the haddocks to `hasDuplicatesConfigPath`.

```haskell
-- | Check if the path has duplicates. A cycle of imports is not allowed. This
-- check should only be done after the path has been canonicalized with
-- @canonicalizeConfigPath@. This is because the import path may contain paths
-- that are the same in relation to their importers but different in relation to
-- the project root.
hasDuplicatesConfigPath :: ProjectConfigPath -> Bool
hasDuplicatesConfigPath (ProjectConfigPath p) = length p /= length (NE.nub p)
```

Checking ``importLoc `elem` seenImports`` may have worked if there was a time when all imports were relative to the root project. In any case, I added tests with #9680 that showed that it wasn't enough as it was. The failures can be seen in `cabal-testsuite/PackageTests/ConditionalAndImport/cabal.test.hs`.

I added more tests for cyclical imports with #9665. Before that the only test we had for cycles was that a project didn't import itself, with `cabal-testsuite/PackageTests/ConditionalAndImport/cabal-cyclical.project`;

```
packages: .

import: cabal-cyclical.project
```

Reply:
I meant to ask why ``normLocPath `elem` normSeenImports`` doesn't catch the circular import, where `normSeenImports` is relative to the project directory. However, I just realized that I don't really understand the reason for `seenImports`/`normSeenImports`, since it contains full `ProjectConfigPath`s. I would have expected the circularity check to check whether the new import (as a normalized `FilePath`) is an element in the current  path of imports (as a list of normalized `FilePath`s or a single normalized `ProjectConfigPath`).

Reply:
The `seenImports` only ever act like a stack. I've just added a test using `cabal-testsuite/PackageTests/ConditionalAndImport/yops-0.project` that checks to see if we can detect multiple paths to the same `.config` file (so effectively the same file imported multiple times) and we can't yet. I'm worried that I may have to reach for an `MVar`.

Reply:
I was able to catch the duplicate import in the "Y-shaped" `yops` test using an `IORef [FilePath]`.

Reply:
Here's the check now;

```haskell
if hasDuplicatesConfigPath normLocPath || uniqueImport `elem` seenUniqueImports
```

The first checks for simple cycles while the second checks for the "Y-shaped" duplicates.

Reply:
I didn't realize that importing the same config file twice was a problem. If so, shouldn't it have a different error message?